src/js/shop/components/Checkout/Checkout.js

```diff
@@ -0,0 +1,259 @@
+        "navigation__link--enabled": enabled,
+    });
+
+const camelize = (obj) => {
+    // recurse into arrays
+    if (Array.isArray(obj)) {
+        return obj.map(camelize);
+    }
+
+    if (!isObject(obj)) {
+        return obj;
+    }
+
+    // convert keys to camelCase
+    const newObj = {};
+    Object.entries(obj).forEach(([key, value]) => {
+        const newKey = camelCase(key);
+        const newValue = camelize(value);
+        newObj[newKey] = newValue;
+    });
+
+    return newObj;
+};
+
+const NavLink = ({ enabled = false, className, ...props }) => {
+    const Container = enabled ? RRNavLink : "span";
+    const wrappedClassname = enabled
+        ? ({ isActive }) => className({ isActive, enabled })
+        : className({ isActive: false, enabled });
+    return <Container {...props} className={wrappedClassname} />;
+};
+
+const initialState = {
+    checkoutMode: "withoutAccount",
+    customer: {
+        firstName: "",
+        lastName: "",
+        email: "",
+        phone: "",
+    },
+    deliveryAddress: EMPTY_ADDRESS,
+    billingAddress: null, // same as delivery address
+    addressStepValid: false,
+};
+
+const reducer = (draft, action) => {
+    switch (action.type) {
+        case "FIELD_CHANGED": {
+            const { name, value } = action.payload;
+            set(draft, name, value);
+            break;
+        }
+        case "CHECK_ADDRESS_VALIDITY": {
+            draft.addressStepValid = checkAddressFieldsComplete(
+                draft.customer,
+                draft.deliveryAddress
+            );
+            break;
+        }
+        default: {
+            throw new Error(`Unknown action type: ${action.type}`);
+        }
+    }
+};
+
+/**
+ *
+ * Checkout
+ *
+ */
+const Checkout = ({ cartStore, user }) => {
+    const location = useLocation();
+    const navigate = useNavigate();
+    const checkoutRoot = useHref("/");
+
+    const isAuthenticated = Object.keys(user).length > 1;
+    const customer = camelize(user);
+
+    const dynamicInitialState = {
+        ...initialState,
+        checkoutMode: isAuthenticated ? "withAccount" : "withoutAccount",
+    };
+    if (isAuthenticated) {
+        dynamicInitialState.customer = customer;
+        dynamicInitialState.deliveryAddress = {
+            ...dynamicInitialState.deliveryAddress,
+            ...customer.profile,
+        };
+        dynamicInitialState.deliveryAddress.postalCode =
+            dynamicInitialState.deliveryAddress.postal;
+        if (dynamicInitialState.deliveryAddress.country == null) {
+            dynamicInitialState.deliveryAddress.country = "N";
+        }
+    }
+
+    const [state, dispatch] = useImmerReducer(reducer, dynamicInitialState);
+
+    // redirect if on the homepage
+    useEffect(() => {
+        if (location.pathname !== "/") {
+            dispatch({ type: "CHECK_ADDRESS_VALIDITY" });
+        }
+    }, [isAuthenticated, location, navigate, dispatch]);
```

Comment:
```suggestion
    }, [isAuthenticated, location]);
```
`dispatch` should have a stable identity (at least in React it does) so it shouldn't b necessary as dependency. `navigate` isn't used here at all, so prob not needed either.

Reply:
I know - but my experience from eslint plugins is that they complain about missing dependencies while they are used in the hook body. If they're invariants, it doesn't hurt to have them here and I prefer being explicit about them.

Reply:
If you're using correct plugins they'd also complain about extra dependencies. Possible con of extra dependencies is that they can make useEffect's callback be called when you don't expect it.